### PR TITLE
Add multi-choice dialog and prompt dialog

### PR DESCRIPTION
**Pull request by john0312**

### CODE REVIEW

File: common/extlib/client-extension-helper.mjs [hunk 1]
```diff
@@ -64,29 +64,10 @@ class ClientExtensionHelper {
...
-      };
-      // TODO: Handle timeout.
-      this.socket.emit('callC2sAPI', callArgs, (result) => {
-        // TODO: Resolve the promise and return the result.
-        clearTimeout(timeoutTimer);
-        resolve(result);
-      });
-    });
-    return await resultPromise;
+    return this.extMan.callC2sAPI(extName, methodName, timeout, ...args);
```
Comment (ktpss95112) [inline common/extlib/client-extension-helper.mjs:70]: Should be `return await callC2sAPI(...)`?

File: common/extlib/client-extension-manager.mjs [hunk 2]
```diff
@@ -130,10 +130,45 @@ class ClientExtensionManager {
       return await this.extHelpers[extName].onS2cAPICalled(methodName, args);
     } catch (e) {
       console.error(`Exception '${e}' calling S2cAPI '${methodName}' in '${extName}'`);
+      console.log(e.stack);
```
Comment (ktpss95112) [inline common/extlib/client-extension-manager.mjs:133]: `console.trace()`

File: common/extlib/client-extension-manager.mjs [hunk 3]
```diff
@@ -130,10 +130,45 @@ class ClientExtensionManager {
...
+   * @param {string} extName - The name of the extension.
+   * @param {string} methodName - The name of the API.
+   * @param {Number} timeout - An optional timeout in ms.
+   * @param {Array} args - Arguments to the call.
+   * @return {object} result - The result from the call.
+   */
+  async callC2sAPI(extName, methodName, timeout, ...args) {
+    // TODO: Emit the corresponding event through socket in game client.
+    if (!timeout) timeout = 0;
+    console.assert(extName, 'extName in callC2sAPI() should be valid');
```
Comment (ktpss95112) [inline common/extlib/client-extension-manager.mjs:150]: The functionality of `console.assert(extName)` seems strange ...

File: common/extlib/client-extension-manager.mjs [hunk 4]
```diff
@@ -130,10 +130,45 @@ class ClientExtensionManager {
...
+    const resultPromise = new Promise((resolve) => {
+      // TODO: Fill in callArgs so gateway service knows how to handle it.
+      const timeoutTimer = setTimeout(() => {
+        resolve({error: 'timeout'});
+      }, timeout);
+
+      let callArgs = {
```
Comment (ktpss95112) [inline common/extlib/client-extension-manager.mjs:157]: Use `const`?

File: extensions/dialog/common/client.mjs [hunk 5]
```diff
@@ -15,42 +15,113 @@ class DialogModal extends Modal {
...
     this.subjectDom = document.getElementById(DIALOG_SUBJECT_DIV);
-    this.btnContDom = document.getElementById(DIALOG_BUTTON_CONTAINER_DIV);    
+    this.btnContDom = document.getElementById(DIALOG_BUTTON_CONTAINER_DIV);
   }
 
-  /**
-   * Call to set the dialog to display a single choice dialog.
-   */
-  async displayAsSingleChoice(subject, message, buttonText) {
+  async displayDialogInternal(setupDom) {
```
Comment (ktpss95112) [inline extensions/dialog/common/client.mjs:21]: If this method is not designed for external calling, maybe prepend an underscore: `async _displayDialogInternal(setupDom) {`

File: extensions/dialog/common/client.mjs [hunk 6]
```diff
@@ -15,42 +15,113 @@ class DialogModal extends Modal {
...
+      const promptHtml = '<br /><input type="text" class="dialog-textinput"' +
+        ' id="dialog-textinput" />';
+      this.msgDom.innerHTML = message + promptHtml;
+      this.btnContDom.innerHTML = `<button id="dialog-btn-OK">${buttonText}</button>`;
+      const btnOK = document.getElementById('dialog-btn-OK');
+
+      // Set the visibilities right.
+      this.btnContDom.classList.add('visible');
+
+      return [btnOK];
```
Comment (ktpss95112) [inline extensions/dialog/common/client.mjs:118]: Why return an `Array`?

**Review by ktpss95112 — COMMENTED** [review]
add some comments

File: extensions/dialog/common/client.mjs (hunk 6 above)
Comment (john0312) [reply extensions/dialog/common/client.mjs:118]: As mentioned, it's a callback that expects array for return.

File: common/extlib/client-extension-manager.mjs [hunk 7]
```diff
@@ -130,10 +130,46 @@ class ClientExtensionManager {
...
+  /**
+   * Call an API provided by the same extension's standalone part
+   * on the server side.
+   * @param {string} extName - The name of the extension.
+   * @param {string} methodName - The name of the API.
+   * @param {Number} timeout - An optional timeout in ms.
+   * @param {Array} args - Arguments to the call.
+   * @return {object} result - The result from the call.
+   */
+  async callC2sAPI(extName, methodName, timeout, ...args) {
```
Comment (ktpss95112) [inline common/extlib/client-extension-manager.mjs:147]: Maybe support `Infinity` timeout or call API without timeout?

**Pull request merged**